shell/platform/android/io/flutter/embedding/android/FlutterView.java

```diff
@@ -508,30 +509,50 @@ private int guessBottomKeyboardInset(WindowInsets insets) {
   public final WindowInsets onApplyWindowInsets(@NonNull WindowInsets insets) {
     WindowInsets newInsets = super.onApplyWindowInsets(insets);
 
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+      Insets systemGestureInsets = insets.getSystemGestureInsets();
+      viewportMetrics.systemGestureInsetTop = systemGestureInsets.top;
+      viewportMetrics.systemGestureInsetRight = systemGestureInsets.right;
+      viewportMetrics.systemGestureInsetBottom = systemGestureInsets.bottom;
+      viewportMetrics.systemGestureInsetLeft = systemGestureInsets.left;
+    }
+
     boolean statusBarVisible = (SYSTEM_UI_FLAG_FULLSCREEN & getWindowSystemUiVisibility()) == 0;
     boolean navigationBarVisible =
         (SYSTEM_UI_FLAG_HIDE_NAVIGATION & getWindowSystemUiVisibility()) == 0;
 
     if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
-      int mask = 0;
+      int mask = android.view.WindowInsets.Type.ime();
       if (navigationBarVisible) {
         mask = mask | android.view.WindowInsets.Type.navigationBars();
       }
       if (statusBarVisible) {
         mask = mask | android.view.WindowInsets.Type.statusBars();
       }
-      mask = mask | android.view.WindowInsets.Type.ime();
+      Insets uiInsets = insets.getInsets(mask);
 
-      Insets finalInsets = insets.getInsets(mask);
-      viewportMetrics.paddingTop = finalInsets.top;
-      viewportMetrics.paddingRight = finalInsets.right;
+      viewportMetrics.paddingTop = uiInsets.top;
+      viewportMetrics.paddingRight = uiInsets.right;
       viewportMetrics.paddingBottom = 0;
-      viewportMetrics.paddingLeft = finalInsets.left;
+      viewportMetrics.paddingLeft = uiInsets.left;
 
       viewportMetrics.viewInsetTop = 0;
```

Comment:
Should we getInsets multiple times? Since we're feeding them into different things.

Reply:
Yes, I'll take this opportunity to fully use the new API rather than just adapt the old version.